Added sampling methods of crv_types

Smit-create opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->
Added sampling methods of continuous random variables
#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Closes #17057 
Related to #19061
#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* stats
   * Added sampling methods for continuous variables
   * Added `library` option in `sample`
   *  `sample` returns an iterator object since version 1.7
<!-- END RELEASE NOTES -->

Smit-create committed:
added sampling of crv

Smit-create committed:
corrected tests

Smit-create committed:
corrected tests

Smit-create committed:
code quality

Smit-create commented:
Please review.

Smit-create committed:
size correction in python

Smit-create commented:
This is ready for review. @czgdp1807

Smit-create committed:
space correction

czgdp1807 commented on sympy/stats/crv.py:
```diff
@@ -173,11 +176,26 @@ def check(*args):
         if not size:
-            return icdf(random.uniform(0, 1))
+            size = 1
+        if getattr(self,'_sample_python', None):
```
Why not `hasattr`?

Smit-create replied:
Yes, `hasattr` would be good to use here. Thanks!

czgdp1807 commented:
The diff coverage is quite low. Can you add more tests or modify current tests to or merge master to increase the coverage?

Smit-create committed:
code cov

Smit-create commented:
![Screenshot from 2020-03-04 16-21-20](https://user-images.githubusercontent.com/55887635/75872564-68560680-5e34-11ea-962a-9a12a20b4c46.png)
Code coverage now increases to `100%`

Smit-create commented:
@czgdp1807  Does it look good to go?

czgdp1807 commented:
LGTM. Will merge if no one raises objections.

czgdp1807 commented on sympy/stats/crv.py:
```diff
@@ -173,11 +176,26 @@ def check(*args):
+            return samp if size != 1 else samp[0]
+        elif hasattr(self,'_sample_pymc3') and import_module('pymc3'):
+            samp = self._sample_pymc3(size)
+            return samp if size != 1 else samp[0]
```
What we if we want to add more libraries in future, like tensorflow? 
Can we generalise this using a loop and storing the library names in a list? I hope you are getting my point.

Smit-create replied:
I think you mean that will this be able to extend the `if, elif` ladder by adding such libraries to this ladder and its corresponding `_sample_library` in its class. According to the current API design that if someone needs to add sampling methods to this from any library, add the library to the ladder, then corresponding method to the class and will work.

Smit-create replied:
```
        if hasattr(self, '_sample_python'):
            samp = self._sample_python(size)
            return samp if size != 1 else samp[0]

        libraries = ['numpy', 'scipy', 'pymc3']
        for lib in libraries:
            if hasattr(self, '_sample_' + lib ) and import_module(lib):
                samp_method = getattr(self, '_sample_' + lib)
                samp = samp_method(size)
                return samp if size != 1 else samp[0]
```
Do you mean this way? This could also be used. As when in future we add more libraries, we just need to add the library name in list and its corresponding method in the class. Does this sound good?

czgdp1807 replied:
Yeah, this kind of generalisation will be better.

Smit-create replied:
Okay, then I will change accordingly

czgdp1807 commented on sympy/stats/crv_types.py:
```diff
@@ -364,11 +370,28 @@ def pdf(self, x):
+        a, b = float(self.alpha), float(self.beta)
+        with pymc3.Model():
+            pymc3.Beta('X', alpha=a, beta=b)
+            return pymc3.sample(size, chains=1, progressbar=False)[:]['X']
```
Similarly more methods will have to be added for new libraries. Can we generalise this too?

Smit-create replied:
I think we can generalize this as this becomes easy to continue according to current API design in the future. But still, I will look into it in more detail if I could come up with some other better way.

oscarbenjamin commented:
The code here is very repetitive. It would be better to make a data structure that stores the different functions like in lambdify and the code printers.

Would it make more sense to handle this as part of lambdify and the code printers?

Smit-create commented:
> Would it make more sense to handle this as part of lambdify and the code printers?

I will look into this suggestion.

Smit-create committed:
remove if elif

oscarbenjamin commented:
It probably makes more sense to group the code together in one class for each of the libraries and have a mapping from sympy types to the other library types rather than having each of the sympy classes repeat the boiler plate for calling into the external library. That is what I mean by saying that it is repetitive. If we have  m other libraries and k sympy distribution classes then we will end up having to add `m*k` methods for connecting them all. Since m is smaller than k it makes more sense to have m classes for the output each of which can have a mapping of size k that efficiently describes how to connect between them.

Smit-create commented:
Thanks! @oscarbenjamin, I get your point. I will think of some way to work upon that.

Smit-create commented:
I have designed another way. I will explain with an example of `sample_scipy`.

We will create 5 dictionaries, one which will import the corresponding random variable from scipy, other four will map scipy arguments with sympy arguments. Generally, scipy has 4 main arguments, i.e, `a` , `b`, `loc`, `scale`. The other four dictionaries will map these arguments using attributes of the corresponding class of random variable. 
Then, finally, return the sample using mapped random variable from first dictionary and arguments from other 4 dictionaries.

Smit-create commented:
I have designed a function that works fine with scipy:
```

def _sample_scipy(dist, size):

    dist_list =  ['BetaDistribution', 'BetaPrimeDistribution',
    'CauchyDistribution', 'ChiDistribution', 'ChiSquaredDistribution',
    'ExponentialDistribution', 'GammaDistribution', 'GammaInverseDistribution',
    'LogNormalDistribution', 'NormalDistribution', 'GaussianInverseDistribution',
    'ParetoDistribution', 'UniformDistribution']

    if dist.__class__.__name__ not in dist_list:
        return None
    from scipy.stats import (beta, betaprime, cauchy, chi, chi2, expon, gamma,
                             invgamma, lognorm, norm, invgauss, pareto, uniform)

    scipy_rv_map = {
            'BetaDistribution': lambda dist, size: beta.rvs(a=float(dist.a), b=float(dist.b),
                                    size=size),
            'BetaPrimeDistribution':lambda dist, size: betaprime.rvs(a=float(dist.alpha),
                                    b=float(dist.beta), size=size),
            'CauchyDistribution': lambda dist, size: cauchy.rvs(loc=float(dist.x0),
                                    scale=float(dist.gamma), size=size),
            'ChiDistribution': lambda dist, size: chi.rvs(df=float(dist.k), size=size),
            'ChiSquaredDistribution': lambda dist, size: chi2.rvs(df=float(dist.k), size=size),
            'ExponentialDistribution': lambda dist, size: expon.rvs(loc=0, scale=1/float(dist.rate),
                                    size=size),
            'GammaDistribution': lambda dist, size: gamma.rvs(a=float(dist.k), loc=0,
                                    scale=float(dist.theta), size=size),
            'GammaInverseDistribution': lambda dist, size: invgamma.rvs(a=float(dist.a), loc=0,
                                    scale=float(dist.b), size=size),
            'LogNormalDistribution': lambda dist, size: lognorm.rvs(s=std, loc=0,
                                    scale=exp(float(dist.mean), size=size)),
            'NormalDistribution': lambda dist, size: norm.rvs(float(dist.mean), float(dist.std), size=size),
            'GaussianInverseDistribution': lambda dist, size: invgauss.rvs(
                mu=float(dist.mean)/float(dist.shape), scale=float(dist.shape), size=size),
            'ParetoDistribution': lambda dist, size: pareto.rvs(b=float(dist.alpha),
                                    scale=float(dist.xm), size=size),
            'UniformDistribution': lambda dist, size: uniform.rvs(loc=float(dist.left),
                            scale=float(dist.right)-float(dist.left), size=size),
        }
    return scipy_rv_map[dist.__class__.__name__](dist, size)
```

This will remove all `_sample_scipy` methods from individual classes, and just a single function will be used.
@oscarbenjamin @czgdp1807 Does this look good to commit for all libraries?

oscarbenjamin commented:
That looks better. There's probably a way to organise this nicely into one class for each external library and perhaps factor some code out to a base class.

Smit-create committed:
changed sampling design

Smit-create committed:
remove unused imports

Smit-create committed:
import random in sample_python

Smit-create committed:
remove random import

Smit-create committed:
fix failing tests

Smit-create commented:
@czgdp1807  @oscarbenjamin  Please review this

Smit-create commented:
@czgdp1807 @oscarbenjamin Does this look good?

oscarbenjamin commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -1497,3 +1496,92 @@ def test_ContinuousDistributionHandmade():
+
+    numpy = import_module('numpy')
+    if not numpy:
+        skip('Numpy is not installed. Abort tests for _sample_numpy.')
```
This will skip all the tests below as well

Smit-create replied:
So should I break this into three different tests?

oscarbenjamin commented:
I think this looks better organised

Smit-create committed:
Merge branch 'master' into sample_crv

Smit-create committed:
break tests

Smit-create committed:
increase coverage

Smit-create commented:
I have added the tests of each sampling method separately to increase the diff coverage. Does this look ready to merge? @czgdp1807

oscarbenjamin commented:
I don't know the stats code so well myself but this seems better organised than the diff before.

Smit-create commented:
@czgdp1807 Please review this PR.

Smit-create commented:
ping @czgdp1807

sylee957 commented:
In my opinion, I don't think that it is just better organized than before because
1. Why does it have to generate all the dictionaries inside the calls?
2. Why should a new object created with attributes like `distribution`, `size` if they are only used once?

I think that if the objects are just used once and thrown away, there's no reason to have the attributes but a collection of classmethods.

```
class Sampler:
    scipy_sampler_dict
    numpy_sampler_dict
    ...

    @classmethod
    def sample_scipy(cls, dist, size):
         ... 
```

However, I see that scipy samplers classes store some state. So if you are making a new object for sampler, you should take account for that.

Smit-create commented:
@sylee957  Thanks for you suggestion. I will surely make the required changes and make this look clearer and simpler to be extended further.

Smit-create committed:
changed to classmethod

Smit-create committed:
Merge branch 'master' into sample_crv

Smit-create commented:
@sylee957 @czgdp1807 How does this look now? In my opinion, this one is precise, easy to understand and extend. Please share your views on this.

sylee957 commented:
I see there are some occurances of `_sample_scipy`, `_sample_pymc3` remaining in `drv` or `frv`.

Smit-create commented:
Yes, once the sampling API is finalized and merged for continuous. I will make corresponding changes in sampling of discrete and finite RVs.

sylee957 commented on sympy/stats/crv.py:
```diff
@@ -145,6 +150,164 @@ def __call__(self, *args):
+            handmade_pdf = dist.pdf
+            class scipy_pdf(rv_continuous):
+                def _pdf(self, z):
+                    return handmade_pdf(z)
```
How is it supposed to operate?
I think that sympy functions should be lambdified before passing to scipy.

sylee957 commented on sympy/stats/crv.py:
```diff
@@ -173,8 +336,21 @@ def check(*args):
+        if samps is not None:
+            return samps if size!=1 else samps[0]
+
+        libraries = ['scipy', 'numpy', 'pymc3']
```
I think that it can get ambiguous which library it is using for sampling.
I see it already return different types of arrays.

So I think that it should better be specified as an option and unsupport stuff that are not available for specific libraries, than trying all the libraries to get it succeed.

Smit-create replied:
In my opinion, at the user end, we don't need to specify which library is used to sample. Also, for coverage purposes, we can do by directly calling the concerned method from class `SampleExternal`. IMO, We should not go for unsupported kind of stuff, as providing such an option, also will lead to unnecessary raising errors, instead, I suggest just return the sample without any specification of the library. To maintain the uniformity, I will type-cast each of the samples(in case more than one sample is required) returned by different libraries to `list`. Please share your thoughts @czgdp1807.

czgdp1807 replied:
I think, there should be an option provided to the user for using libraries. In this way user will be knowing the type of the output they will be getting. We should keep the public API unambiguous. Type casting is not recommended when we can give an option and make the things easy.

Smit-create replied:
Okay. I will go with the suggestion. As there is a change in the public API of the stats module, I will initially work for the continuous RV.
* Firstly, it will require addition of an argument in each function call of `sample`.
* I will be keeping this as`None` by default for all the calls not used in Continuous RV.
After this gets finalized for Continuous RV, I will open a new PR for the same changes in Finite and Discrete RV, and will make the actual use of lib argument in their function calls also, so that the diff size remains under some bound and becomes easy to be reviewed in parts.
Does this plan look fine?

czgdp1807 replied:
What used to happen earlier before this framework was introduced?
May be to avoid change in the public API, library can be passed as keyword argument and if nothing is given as library then inbuilt python sampling features can be used.

Smit-create replied:
Yes, I meant the same. There was no library argument before, so, by default will be python. But, I wanted to ask that firstly, should I implement this argument usage for Continuous RV, and then make this use of argument for Finite and Discrete?

czgdp1807 replied:
This PR is concerned with CRV, so first we should try out the approach with CRVs and then we can proceed with  FRV and DRV.

Smit-create replied:
Thanks! I will make the necessary changes.

Smit-create committed:
use lambdify

Smit-create committed:
fix define error

sylee957 commented on sympy/stats/crv.py:
```diff
@@ -145,6 +151,165 @@ def __call__(self, *args):
+
+        if dist.__class__.__name__ == 'ContinuousDistributionHandmade':
+            from scipy.stats import rv_continuous
+            z = Symbol('z')
```
Can you use Dummy in place of symbol?

Smit-create replied:
Yes, I need to change this to Dummy.

Smit-create committed:
use Dummy

Smit-create commented:
Any comments on this? @czgdp1807

czgdp1807 commented:
I will try to comment by tomorrow.

Smit-create commented:
Not a problem!

Smit-create committed:
Merge branch 'master' into sample_crv

czgdp1807 commented on sympy/stats/crv.py:
```diff
@@ -173,8 +338,21 @@ def check(*args):
 
     def sample(self, size=()):
```
```suggestion
    def sample(self, size=(), lib=None):
```

czgdp1807 commented on sympy/stats/crv.py:
```diff
@@ -173,8 +338,21 @@ def check(*args):
+                if samps is not None:
+                    return list(samps) if size != 1 else samps[0]
+
+        icdf = self._inverse_cdf_expression()
```
```suggestion
        if import_module(library):
            samps = getattr(SampleExternal, '_sample_' + lib)(self, size)
            if samps is not None:
                return list(samps) if size != 1 else samps[0]

    icdf = self._inverse_cdf_expression()
```

Smit-create committed:
api change: include library option

Smit-create commented:
In the current commit, the default library is kept as `'python'`. IMO, we should by default provide an option as `'any'` which means a user is only interested to have the sample no matter from which library. In this case, we need to iterate over `libraries` as previously. Please provide your opinion on this. @czgdp1807 @sylee957

sylee957 commented:
I don’t think that it’s useful to just thread over all available libraries and see if something is available or not.

The only problem is that some distributions are not available for some libraries. However, I think that it would be better to handcraft some distributions in sympy automatically for them.

Smit-create commented:
There is a failing test on Travis.
```
________________________________________________________________________________

______________ sympy/series/tests/test_limits.py:test_issue_14811 ______________

Traceback (most recent call last):

  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/sympy-1.6.dev0-py3.5.egg/sympy/core/assumptions.py", line 260, in getit

    return self._assumptions[fact]

KeyError: 'zero'
```

Smit-create commented:
> I don’t think that it’s useful to just thread over all available libraries

So, in you opinion, the current diff suffices your suggestion?

sylee957 commented:
Yes, as long as this keeps the behavior as before but is about refactory. The option for ‘any’ can be discussed later.

Smit-create commented:
> Yes, as long as this keeps the behavior as before but is about refactory

Thanks, this would also work fine as the `else` block of `if library == 'python':` almost handles all the cases when all of the four things fails to produce the result.
@czgdp1807 Please share your thoughts on this.

czgdp1807 commented:
IMO, as I said earlier, options like `any` shouldn't be considered as they will lead to an ambiguous API making things difficult for the user. Some libraries, for example TF returns tensors and NumPy returns arrays, so if we will check all libraries for sampling, the output type can be non-deterministic. Or simply if the list of libraries is `[tensorflow, numpy]`. Now in my system, I don't have tensorflow, so sampling will be done through numpy, and in another system, I have tensorflow so sampling will be done through tensorflow. So if I write code using these APIs then either I have to consider all cases(as I don't know what will I get) making life difficult or I will just write for the default python case and the library tensorflow so the code will be cross compatible over various systems.

sylee957 commented on sympy/stats/rv.py:
```diff
@@ -1032,15 +1049,20 @@ def sample(expr, condition=None, size=(), **kwargs):
 
     """
-    return next(sample_iter(expr, condition, size=size, numsamples=1))
+    return next(sample_iter(expr, condition, size=size, library = library,
```
```suggestion
    return next(sample_iter(expr, condition, size=size, library=library,
```

Smit-create commented:
Okay!, So according to you the current diff is the required API we aim for?

Smit-create committed:
code quality

sylee957 commented on sympy/stats/rv.py:
```diff
@@ -1221,7 +1247,7 @@ def sampling_P(condition, given_condition=None, numsamples=1,
     count_false = 0
 
-    samples = sample_iter(condition, given_condition,
+    samples = sample_iter(condition, given_condition, size=size, library=library,
```
I see you have added `size` argument to some functions.
But does the array shape make sense for mean, variance if numsamples is already used?

Smit-create replied:
The actual problem, I see is that is `size` and `numsamples` mean the same, as before, `numsamples` was not provided in public API of `sample`, and so, `size` was introduced. So, we need to take a decision to use either `size` or `numsamples`. We need to raise Deprecation warning in either of the case.

Smit-create replied:
IMO, we should remove `size`, and use `numsamples` as it was been there for a long period of time.

Smit-create replied:
I observed many failing things:
```
>>> from sympy.stats import *
>>> X=Normal('X',1,2)
>>> sample(X)     #works fine
-0.395115576398039

>>> sample(X, size=3)     # works fine
[3.04853254256568, -5.15781049701721, -0.916547668739276]
>>> Y=Normal('Y',3,4)
>>> sample(X+Y, size=3)   # failed : expected 3
2.94035736693013

>>> sample(X+Y, numsamples=3)   # failed : expected 3
5.91353130029414

>>> # adding numsamples as argument in sample function
>>> sample(X+Y, numsamples=3)   # works fine
[2.60480735236754, -2.59005952144665, 1.87621810047909]

>>> sample(X+Y, size=3)   # failed : expected 3
[1.22518985609606]

>>> sample(X+Y, numsamples=3, library='scipy') # fails : expected to produce array output
[1.18343224436589, 5.09387064093793, 5.55350069910508]

>>> sample(X, numsamples=3, library='scipy') # fails : expected to produce array output
[3.9671541494809617, -2.810799332922929, -1.0880356350959333]

>>> sample(X, size=3, library='scipy')  # failed : expected array instead of list
[array([ 2.02042956, -0.71548653,  2.01273328])]
```

czgdp1807 replied:
Looks quite inconsistent. What is the difference between, `numsamples` and `size`?

czgdp1807 replied:
What is used in public API, `size` or `numsamples`?

Smit-create replied:
Actually, `numsamples` has previously used for signifying total number of samples, but the user cannot provide it while calling` sample()`(as shown 4th example which doesn't affect currenlty as is overwritten), therefore `size` was introduced for the same use, but now I see that `size` and `numsamples` both mean the same.

Smit-create replied:
So, I suggest removing `size` and use `numsamples` as it will maintain unifromity across all other samplings in `rv.py` and avoid confusion

czgdp1807 replied:
So, now `size` is what public API has, so `numsamples` should be removed/renamed to `size`.

Smit-create replied:
Some functions like `sampling_P`, `sampling_density`, etc. have `numspaces`

Smit-create replied:
https://github.com/sympy/sympy/blob/63ab534ebc9bb8fa47ca80aea5a1b0eca743fb9e/sympy/stats/rv.py#L1020-L1040
In `sample` we have `size` as public API, while returning `numsaples` is by default kept as 1.

The following samplings have `size` and `numsamples` both as the argument.
https://github.com/sympy/sympy/blob/63ab534ebc9bb8fa47ca80aea5a1b0eca743fb9e/sympy/stats/rv.py#L1043

https://github.com/sympy/sympy/blob/63ab534ebc9bb8fa47ca80aea5a1b0eca743fb9e/sympy/stats/rv.py#L1130

https://github.com/sympy/sympy/blob/63ab534ebc9bb8fa47ca80aea5a1b0eca743fb9e/sympy/stats/rv.py#L1177


The following samplings have only `numsamples` as argument.
https://github.com/sympy/sympy/blob/63ab534ebc9bb8fa47ca80aea5a1b0eca743fb9e/sympy/stats/rv.py#L1207

https://github.com/sympy/sympy/blob/63ab534ebc9bb8fa47ca80aea5a1b0eca743fb9e/sympy/stats/rv.py#L1243

https://github.com/sympy/sympy/blob/63ab534ebc9bb8fa47ca80aea5a1b0eca743fb9e/sympy/stats/rv.py#L1265

Smit-create replied:
@czgdp1807 Please share your thoughts on removing which of this argument, to make it uniform throughout all these functions.

czgdp1807 replied:
What happens if we just let the sampling happen by using default values of `numsamples` and `size` in the above functions? 
The removal should be done in a different PR. Can you open an issue for it as it is concerned with the changes in public API.

czgdp1807 commented on sympy/stats/crv.py:
```diff
@@ -171,13 +336,35 @@ def __new__(cls, *args):
+                else:
+                    return [icdf(random.uniform(0, 1))]*size
+
+        libraries = ['scipy', 'numpy', 'pymc3']
```
```suggestion
        libraries = ['python', 'scipy', 'numpy', 'pymc3']
```

Smit-create replied:
This fails after adding `'python'` in list, So handled out of the list
```python
>>> from sympy.external import import_module
>>> print(import_module('python'))
None
```

czgdp1807 commented on sympy/stats/crv.py:
```diff
@@ -171,13 +336,35 @@ def __new__(cls, *args):
+                if size == 1:
+                    return icdf(random.uniform(0, 1))
+                else:
+                    return [icdf(random.uniform(0, 1))]*size
```
```suggestion
```

Smit-create replied:
IMO,` 'python' `needs to be handled differently. Because it has an alternative if `SampleExternal` fails for `'python'`

czgdp1807 replied:
Why are using that alternative? Can that alternative be shifted to `_sample_python` so that final result is not `None` for `_sample_python` whatever be the case.

czgdp1807 commented on sympy/stats/crv.py:
```diff
@@ -171,13 +336,35 @@ def __new__(cls, *args):
+
+        libraries = ['scipy', 'numpy', 'pymc3']
+        if library not in libraries:
+            raise NotImplementedError("Sampling from %s is not implemeted."
```
```suggestion
            raise NotImplementedError("Sampling from %s is not supported yet."
```

czgdp1807 commented on sympy/stats/crv.py:
```diff
@@ -171,13 +336,35 @@ def __new__(cls, *args):
+            raise NotImplementedError(
+                    "Sampling for %s is not currently implemented from %s"
+                    % (self.__class__.__name__, library)
+                    )
```
```suggestion
        if not import_module(library):
            raise ValueError("Failed to import %s" % library)

        samps = getattr(SampleExternal, '_sample_' + library)(self, size)
        if samps is not None:
             return samps if size != 1 else samps[0]
        raise NotImplementedError(
                    "Sampling for %s is not currently implemented from %s"
                    % (self.__class__.__name__, library)
                    )
```

czgdp1807 commented on sympy/stats/crv.py:
```diff
@@ -171,13 +336,35 @@ def __new__(cls, *args):
         else:
-            return [icdf(random.uniform(0, 1))]*size
+            raise ValueError("Failed to import %s" % library)
+
```
```suggestion
```
Remove `else` as well.

Smit-create committed:
changes including numsamples and size

Smit-create committed:
fix array errors can intial clean ups

Smit-create commented:
```python
>>> from sympy.stats import *
>>> X=Normal('X',1,2)
>>> next(sample(X))
[0.469738066019727]

>>> next(sample(X, size=3))
[1.03193769686300, 4.37068224926506, 2.25993735941436]

>>> Y=Normal('Y',3,4)
>>> next(sample(X+Y, size=3))
[1.85229046187869, 11.3073651698796, 2.46753484877966]

>>> list(sample(X+Y, numsamples=3))
[[-3.92733920820063], [1.91237869520523], [6.39355454594779]]

>>> list(sample(X+Y, numsamples=3, library='scipy'))
[array([2.26730872]), array([-4.65409755]), array([1.24456059])]

>>> list(sample(X, numsamples=3, library='scipy'))
[array([3.16727285]), array([-1.0967364]), array([-2.74534138])]

>>> next(sample(X, size=3, library='scipy'))
array([3.14564194, 0.78263862, 1.14410964])
```

On the current diff, it is able to produce the expected results for all the examples.

I have modified 1-2 failing tests as I had few doubts regarding handling arrays in sympy:
(i) `ImmutableDenseNDimArray` raises an error when substituted in any expression and then using evalf on it, So needed to  modify this    ` assert variance(X + Y, numsamples=10, evalf=False).has(ImmutableDenseNDimArray)`

(ii) Probably if this supports some operations as in numpy arrays like:
```python
>>> import numpy
>>> import numpy as np
>>> np.array([1,2,3])**2
array([1, 4, 9])
```
Will be able to resolve above error.

Smit-create committed:
remove if else from sample in drv

Smit-create committed:
Merge branch 'master' into sample_crv

czgdp1807 commented:
How `ImmutableDenseNDimArray` came into picture? Are we not using [random module](https://docs.python.org/2.0/lib/module-random.html) for `python`?

Smit-create commented:
> How `ImmutableDenseNDimArray` came into picture?

`random` module generates only one value of the `sample` at a time, so it does not take `size` as an argument. So to return various sizes, we sample from random `size` times, and return this `array` like an object, `ArrayComprehensionMap`. I had doubt that do we have any `array` like objects which can perform similar operations as numpy arrays, so when substituted to original `expr`, does not raise error
https://github.com/sympy/sympy/blob/8a7e6b4c4197e9fef8459ccad732de218b75d448/sympy/stats/rv.py#L1172
https://github.com/sympy/sympy/blob/8a7e6b4c4197e9fef8459ccad732de218b75d448/sympy/stats/rv.py#L1203

czgdp1807 commented on sympy/stats/rv.py:
```diff
@@ -1158,12 +1194,12 @@ def sample_iter_lambdify(expr, condition=None, size=(), numsamples=S.Infinity, *
             if condition:  # Check that these values satisfy the condition
                 gd = given_fn(*args)
-                if gd != True and gd != False:
+                if gd != True or gd != False:
```
If I am not wrong, this will always be `True` whether, `gd` is `True`, `False` or `None`.

czgdp1807 replied:
```suggestion
                if gd is not True or gd is not False:
```
If I am wrong, then this can be done.

Smit-create replied:
Ohh, That was done while checking the error path, I will change that.

czgdp1807 commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -1260,8 +1260,8 @@ def test_prefab_sampling():
         for i in range(niter):
-            assert sample(var) in var.pspace.domain.set
-            samps = sample(var, size=size)
+            assert next(sample(var))[0] in var.pspace.domain.set
```
This will raise an `AssertionError` , which the user won't understand. It'd be better to have some message as well.

czgdp1807 commented:
> I had doubt that do we have any `array` like objects which can perform similar operations as numpy arrays, so when substituted to original `expr`, does not raise error

What about returning `list` object. I think you want to apply `xreplace` to an array to replace each of it's elements. You can do that by using `map` with `list`.

Smit-create commented:
> What about returning `list` object.

This will not work as:
```
>>> l1 = [1, 2 , 3]
>>> l2 = [4, 5, 6]
>>> l1 + l2
[1, 2, 3, 4, 5, 6]
```

Smit-create committed:
sample only from external libraries

Smit-create commented:
**Changes Made**
   * Removed `sample_iter_subs` as it was much slower, and renamed `sample_iter_lambdify` to `sample_iter` which handles everything using `lambdify`.
   * Removed sampling support from sympy and handover all the support to external libraries.
   * Removed `'python'` as an option from `library`.

Smit-create commented on sympy/stats/tests/test_rv.py:
```diff
@@ -180,22 +180,21 @@ def test_H():
 
-    # Make sure this doesn't raise an error
-    E(Sum(1/z**Y, (z, 1, oo)), Y > 2, numsamples=3)
+    raises(TypeError, lambda: E(Sum(1/z**Y, (z, 1, oo)), Y > 2, numsamples=3))
```
How to resolve this error?
```python
>>> E(Sum(1/z**Y, (z, 1, oo)), Y > 2, numsamples=3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/smit/Smit Lunagariya/sympy/sympy/stats/rv.py", line 723, in expectation
    return sampling_E(expr, condition, numsamples=numsamples, evalf=evalf)
  File "/home/smit/Smit Lunagariya/sympy/sympy/stats/rv.py", line 1235, in sampling_E
    numsamples=numsamples, **kwargs))
  File "/home/smit/Smit Lunagariya/sympy/sympy/stats/rv.py", line 1184, in return_generator
    yield fn(*args)
  File "<lambdifygenerated-1>", line 2, in _lambdifygenerated
TypeError: 'float' object cannot be interpreted as an integer
```

czgdp1807 replied:
May be a problem with lambdify?

Smit-create replied:
Yes, its a problem with lambdify. Other than it, does it look good?

czgdp1807 replied:
Have you tried a simpler test? I think you are testing, `sampling_E` here.

Smit-create replied:
Yes, It works good
```python
>>> X = Normal('X', 1, 2)
>>> Y = Gamma('Y', 2, 3)
>>> E(X**2 + Y, numsamples=5)
21.3612416190815
>>> E(X**2 + Y**2, numsamples=5)
63.8733491804054
>>> E(X + Y**2, numsamples=5)
57.9285966139211
>>> E((X + Y)**2, numsamples=5)
86.3801421725767
```

czgdp1807 replied:
You can use the above tests.
Can you just add that test(`E(Sum(1/z**Y, (z, 1, oo)), Y > 2, numsamples=3)`) under allowed failure(use @XFAIL) and fix it in a new a PR later on.

Smit-create committed:
skip if scipy not found

Smit-create committed:
xfail samplingE

Smit-create committed:
Merge branch 'master' into sample_crv

Smit-create committed:
remove docs for deleted functions

Smit-create committed:
typo

Smit-create commented:
Can anyone please restart the build?

Smit-create committed:
use external libraries for sample

Smit-create commented:
Can anyone please have a look at the failing test [https://travis-ci.org/github/sympy/sympy/jobs/677325425](https://travis-ci.org/github/sympy/sympy/jobs/677325425)?

czgdp1807 commented on doc/src/modules/stats.rst:
```diff
@@ -117,8 +117,6 @@ Interface
 .. autofunction:: std
 .. autofunction:: sample
 .. autofunction:: sample_iter
-.. autofunction:: sympy.stats.rv.sample_iter_lambdify
-.. autofunction:: sympy.stats.rv.sample_iter_subs
```
Probably the failure of https://travis-ci.org/github/sympy/sympy/jobs/677325425 is due to this change. Have you tried building the docs locally.
ping @sylee957 Can you please help regarding the SPHINX failure? Thanks.

sylee957 replied:
I think that the HTML build succeeded but have only found the abnormality in doctr. So it can be temporary.
But I think that author needs to resolve the merge conflicts to see if it works for an another trial.

Smit-create committed:
Merge branch 'master' into sample_crv

Smit-create committed:
remove import_module

Smit-create commented:
Please review this.

czgdp1807 commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -1514,6 +1522,87 @@ def test_ContinuousDistributionHandmade():
+    ]
+    size = 3
+    numsamples = 5
+    scipy = import_module('scipy')
+    if not scipy:
+        skip('Scipy is not installed. Abort tests for _sample_scipy.')
```
Shift this to the entry point of the function. No point in creating objects if the module isn't imported.

czgdp1807 commented:
LGTM. Ready for merge.

Smit-create committed:
skip at function start

Smit-create commented:
@czgdp1807  I think we can merge this if no has objection?

czgdp1807 commented:
Let's merge this after, `1.6` branch is created. Some public API change is made in this PR, and various changes made in the coming months to the sampling APIs, so 1.7 will be more appropriate to have this change.

czgdp1807 commented on sympy/stats/rv.py:
```diff
@@ -1017,69 +1021,52 @@ def where(condition, given_condition=None, **kwargs):
+        by default is 'scipy'
+    numsamples : int
+        Number of samples, each with size as ``size``
+
```
Add returns as well here.

czgdp1807 commented on sympy/stats/rv.py:
```diff
@@ -1017,69 +1021,52 @@ def where(condition, given_condition=None, **kwargs):
     >>> N = Normal('N', 3, 4)
-    >>> samp = sample(N)
-    >>> samp in N.pspace.domain.set
+    >>> samp = next(sample(N))[0] # doctest: +SKIP
```
This appears to be a change in public API. Please go through https://github.com/sympy/sympy/wiki/Deprecating-policy and a suitable warning here.

Smit-create replied:
We have deleted the functions `sample_iter_lambdify` and `sample_iter_subs`. Also, in `sample` and `sample_iter`, there is a change in the return type. So I think I should add warnings when someone tries to access `sample_iter_lambdify` and `sample_iter_subs`. Rest is only the return type change?

czgdp1807 replied:
> We have deleted the functions `sample_iter_lambdify` and `sample_iter_subs`

Why? Were these functions available in `sympy.stats` namespace?
@Upabjojr Any thoughts on removing these two functions?

> Rest is only the return type change?

May be, we should warning for these too as the output of public functions are changed.

czgdp1807 replied:
My opinion matches with the following from https://github.com/sympy/sympy/issues/19061#issuecomment-608267737

> However, I think that this doesn't need much deprecation if it works just fine, but there don't have to be any addition for it.

We can keep both the functions and warnings can be added for the change returned types everywhere.

czgdp1807 replied:
See my suggestions below.

Smit-create replied:
> Why? Were these functions available in `sympy.stats` namespace?

No, there were not present in `sympy.stats` namespace.
Also, if they are not used anywhere, and `sample_iter`, now is capable of handling the need using `lambdify`( which was previously `sample_iter_lambdify`) and `sample_iter_subs` is too slow and doesn't follow the discussions in #14261. So, IMHO we should remove these two functions if not now then in further versions.

Smit-create committed:
add deprecated warnings

Smit-create committed:
Merge branch 'master' into sample_crv

czgdp1807 commented on sympy/stats/rv.py:
```diff
@@ -1173,39 +1199,33 @@ def return_generator():
-    sample_iter
+def sample_iter_subs(expr, condition=None, size=(1,), numsamples=S.Infinity,
+                                                                    **kwargs):
+    SymPyDeprecationWarning(
+        feature='sample_iter_subs',
+        useinstead='sample_iter',
+        issue=19061,
+        deprecated_since_version=1.6,
+    ).warn()
```
```suggestion
```

czgdp1807 commented on sympy/stats/rv.py:
```diff
@@ -1173,39 +1199,33 @@ def return_generator():
 
+def sample_iter_lambdify(expr, condition=None, size=(1,), numsamples=S.Infinity,
+                                                                    **kwargs):
+    SymPyDeprecationWarning(
+        feature='sample_iter_lambdify',
+        useinstead='sample_iter',
+        issue=19061,
+        deprecated_since_version=1.6,
+    ).warn()
```
```suggestion
```

czgdp1807 commented on sympy/stats/rv.py:
```diff
@@ -1017,69 +1022,58 @@ def where(condition, given_condition=None, **kwargs):
+    Returns
+    =======
 
+    sample: iterator object
```
I see always an iterator object is returned, so probably a warning can be added for the same.

czgdp1807 replied:
Or adding it in `sample_iter` would be a better idea because every function is using it to generate results, so warning would be raised on any call to it.

Smit-create replied:
We need to add a warning of a return type change in only `sample` as others returned iterator object in current master too.

Smit-create replied:
Should we add something like this?
```python
>>> from sympy.stats import *
>>> X = Normal('X',1, 2)
>>> itr = sample(X)
/home/smit/Smitlunagariya/sympy/sympy/stats/rv.py:1074: UserWarning: The return type of sample has been changed to return an iterator object.
For more information see https://github.com/sympy/sympy/issues/19061
  warnings.warn("The return type of sample has been changed to return an "
>>> next(itr)
array([1.34687082])
```

czgdp1807 replied:
Looks good.

czgdp1807 replied:
`sample returns an iterator object since version 1.7` would be better though.

czgdp1807 replied:
Update the release notes entry as well before merging about this change.

Smit-create replied:
According to the discussions, this looks good to merge now as:
* we have added warnings of change in return type.
* The functions `sample_iter_subs` and `sample_iter_lambdify` are present currently, but I suggest removing them in further versions, as they are of no use now. If you agree with the same then I will open an issue for the removal of them in future

czgdp1807 replied:
> If you agree with the same then I will open an issue for the removal of them in future

That's better. We have a lot of time for `1.7`.

Smit-create committed:
add warnings for sample

czgdp1807 commented:
Finally, it's good to go in. Will merge it by tonight.

Upabjojr commented on sympy/stats/frv.py:
```diff
@@ -339,7 +339,7 @@ def conditional_space(self, condition):
         return FinitePSpace(domain, density)
 
-    def sample(self, size=()):
+    def sample(self, size=(1,), library='scipy'):
```
What is the purpose of substituting `()` with `(1,)` or `1` ? The tuple notation should represent the dimensions of an N-dimensional array you want to be returned.
